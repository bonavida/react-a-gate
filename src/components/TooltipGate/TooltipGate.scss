.tooltip {
  &__container {
    position: absolute;
    display: inline-block;

    &--top {
      margin-top: -5px;
    }

    &--bottom {
      margin-top: 5px;
    }

    &--left {
      margin-left: -5px;
    }

    &--right {
      margin-left: 5px;
    }
  }

  &__arrow {
    position: absolute;
    width: 0;
    height: 0;
    border-style: solid;
    z-index: 3;

    &--top {
      margin-top: -6px;
      border-width: 5px 5px 0;
      border-bottom-color: transparent !important;
      border-left-color: transparent !important;
      border-right-color: transparent !important;
    }

    &--bottom {
      margin-top: 6px;
      border-width: 0 5px 5px;
      border-top-color: transparent !important;
      border-left-color: transparent !important;
      border-right-color: transparent !important;
    }

    &--left {
      margin-left: -6px;
      border-width: 5px 0 5px 5px;
      border-top-color: transparent !important;
      border-bottom-color: transparent !important;
      border-right-color: transparent !important;
    }

    &--right {
      margin-left: 6px;
      border-width: 5px 5px 5px 0;
      border-top-color: transparent !important;
      border-bottom-color: transparent !important;
      border-left-color: transparent !important;
    }
  }

  &__arrow-border {
    z-index: 2;

    &.tooltip__arrow {
      &--top {
        margin-left: -1px;
        border-width: 6px 6px 0;
      }
      &--bottom {
        margin-top: 5px;
        margin-left: -1px;
        border-width: 0 6px 6px;
      }
      &--left {
        margin-top: -1px;
        border-width: 6px 0 6px 6px;
      }
      &--right {
        margin-left: 5px;
        margin-top: -1px;
        border-width: 6px 6px 6px 0;
      }
    }
  }

  &__inner {
    width: max-content;
    max-width: 200px;
    padding: 0.3rem 0.5rem;
    font-size: 0.8em;
    text-align: center;
    border-radius: 0.25rem;
  }
}

.dark {
  .tooltip__inner {
    background-color: #222;
    border: 1px solid #666;
    color: #fff;
    box-shadow: 0 1px 10px rgba(0, 0, 0, 0.3);
  }

  .tooltip__arrow {
    border-color: #222;
  }

  .tooltip__arrow-border {
    border-color: #666;
  }
}

.light {
  .tooltip__inner {
    background-color: #fff;
    border: 1px solid #e1e4e8;
    color: #333;
    box-shadow: 0 8px 24px rgba(149, 157, 165, 0.2);
  }

  .tooltip__arrow {
    border-color: #fff;
  }

  .tooltip__arrow-border {
    border-color: rgba(27, 31, 35, 0.13);
  }
}
