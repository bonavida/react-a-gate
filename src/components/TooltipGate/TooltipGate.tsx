import * as React from 'react';
import { useState, useRef, useCallback, useLayoutEffect } from 'react';
import { createPortal } from 'react-dom';
/** Hooks */
import usePortal from '../../hooks/usePortal';
import useResizeObserver from '../../hooks/useResizeObserver';
/** Utils */
import { calculatePosition } from '../../utils/utils';
/** Types */
import { PositionAttrs, Place, Theme } from '../../types/types';
/** Styles */
import './TooltipGate.scss';

export type TooltipGateProps = {
  rootId?: string;
  className?: string;
  content: string | JSX.Element;
  place?: Place;
  theme?: Theme;
  offset?: number;
};

const TooltipGate: React.FC<TooltipGateProps> = ({
  rootId = 'portal-root',
  className,
  content,
  place = 'top',
  theme = 'dark',
  offset = 10,
  children,
}) => {
  const [showTooltip, setShowTooltip] = useState(false);
  const [coords, setCoords] = useState({ top: -99999, left: -99999 });
  const [arrowCoords, setArrowCords] = useState({ top: -99999, left: -99999 });
  const [placeTooltip, setPlaceTooltip] = useState(place);
  const tooltipRef = useRef<HTMLDivElement>(null);
  const arrowRef = useRef<HTMLDivElement>(null);
  const triggerRef = useRef<JSX.Element>(null);
  const target = usePortal(rootId);

  const updatePosition = useCallback(() => {
    if (triggerRef.current) {
      const {
        top,
        left,
        arrowTop,
        arrowLeft,
        place: newPlace,
      }: PositionAttrs = calculatePosition({
        triggerRef,
        portalRef: tooltipRef,
        arrowRef,
        place,
        offset,
      });
      setCoords({ top, left });
      setArrowCords({ top: arrowTop, left: arrowLeft });
      setPlaceTooltip(newPlace);
    }
  }, [place, offset]);

  useResizeObserver(document.body, updatePosition);

  useLayoutEffect(() => {
    if (tooltipRef.current) {
      updatePosition();
    }
  }, [showTooltip, updatePosition]);

  return (
    <>
      {React.cloneElement(children as JSX.Element, {
        onMouseEnter: () => {
          setShowTooltip(true);
        },
        onMouseLeave: () => {
          setShowTooltip(false);
        },
        'data-testid': 'tooltip_trigger',
        ref: triggerRef,
      })}
      {showTooltip &&
        createPortal(
          <div className={`${theme} ${className || ''}`}>
            <div
              ref={tooltipRef}
              className={`tooltip__container tooltip__container--${placeTooltip}`}
              style={{ ...coords }}
            >
              <div className="tooltip__inner">{content}</div>
            </div>
            <div
              ref={arrowRef}
              className={`tooltip__arrow tooltip__arrow--${placeTooltip}`}
              style={{ ...arrowCoords }}
            />
            <div
              className={`tooltip__arrow tooltip__arrow-border tooltip__arrow--${placeTooltip}`}
              style={{ ...arrowCoords }}
            />
          </div>,
          target
        )}
    </>
  );
};

export default TooltipGate;
