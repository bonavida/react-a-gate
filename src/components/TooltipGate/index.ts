import TooltipGate from './TooltipGate';

export * from './TooltipGate';
export default TooltipGate;
