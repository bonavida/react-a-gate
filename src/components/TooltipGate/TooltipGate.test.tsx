import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
/** Mocks */
import ResizeObserverMock from '../../mocks/ResizeObserver';
/** Components and types */
import TooltipGate, { TooltipGateProps } from './TooltipGate';

afterEach(cleanup);

describe('TooltipGate', () => {
  let props: TooltipGateProps;

  beforeEach(() => {
    props = { content: 'tooltip text' };
    window.ResizeObserver = ResizeObserverMock;
  });

  it('should match snapshot', () => {
    // Arrange
    const { asFragment } = render(
      <TooltipGate {...props}>
        <span>test</span>
      </TooltipGate>
    );

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display trigger content', () => {
    // Arrange
    const { queryByText } = render(
      <TooltipGate {...props}>
        <span>test</span>
      </TooltipGate>
    );

    // Assert
    expect(queryByText('test')).toBeTruthy();
  });

  it('should not display tooltip content if trigger element is not being hovered', () => {
    // Arrange
    const { queryByText } = render(
      <TooltipGate {...props}>
        <span>test</span>
      </TooltipGate>
    );

    // Assert
    expect(queryByText('tooltip')).toBeFalsy();
  });

  it('should display tooltip content if trigger element is being hovered', () => {
    // Arrange
    const { getByText, getByTestId } = render(
      <TooltipGate {...props}>
        <span>test</span>
      </TooltipGate>
    );

    fireEvent.mouseEnter(getByTestId('tooltip_trigger'));

    // Assert
    expect(getByText(/tooltip/i)).toBeTruthy();
  });
});
