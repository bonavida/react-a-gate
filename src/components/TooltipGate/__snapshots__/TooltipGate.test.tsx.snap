// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipGate should match snapshot 1`] = `
<DocumentFragment>
  <span
    data-testid="tooltip_trigger"
  >
    test
  </span>
</DocumentFragment>
`;
